common/src/app/common/logic/tokens.cljc

```diff
@@ -29,14 +29,101 @@
+        from (nth tree from-index)
+        to (nth tree to-index)
+        before (case position
+                 :top to
+                 :bot (nth tree (inc to-index) nil)
+                 :center nil)
+        prev-before (if (:group? from)
+                      (->> (drop (inc from-index) tree)
+                           (filter (fn [element]
+                                     (<= (:depth element) (:depth from))))
+                           (first))
+                      (nth tree (inc from-index) nil))
+
+        drop-as-direct-group-child? (or
+                                     (= :center position)
+                                     (and
+                                      (= :bot position)
+                                      (:group? to)
+                                      (not (get collapsed-paths (:path to)))))
+        from-path (:path from)
+        to-parent-path (if drop-as-direct-group-child?
+                         (:path to)
+                         (into [] (butlast (:path to))))
+        to-path (conj to-parent-path (last from-path))
+
+        identical? (or (= from-index to-index)
+                       (and (= from-path to-path)
+                            (case position
+                              :top (= from-index (dec to-index))
+                              :bot (= from-index to-index)
+                              nil)))
+        to-exists? (and
+                    (not= (:parent-path from) to-parent-path)
+                    (if (:group? from)
+                      (ctob/set-group-path-exists? tokens-lib to-path)
+                      (ctob/set-path-exists? tokens-lib to-path)))
+        parent-to-child-drop? (and
+                               (not= (:parent-path from) to-parent-path)
+                               (:group? from)
+                               (vec-starts-with? to-path (:path from)))]
+    (cond
+      identical? nil
+      to-exists?
+      (throw (ex-info "move token set error: path exists"
+                      {:error :path-exists
+                       :path to-path}))
+      parent-to-child-drop?
+      (throw (ex-info "move token set error: parent-to-child"
+                      {:error :parent-to-child
+                       :from-path from-path
+                       :to-path to-path}))
+      :else
+      (cond-> {:from-path from-path
+               :to-path to-path
+               :before-path nil
+               :before-group? nil}
+        before (assoc :before-path (:path before)
+                      :before-group? (:group? before))
+        prev-before (assoc :prev-before-path (:path prev-before)
+                           :prev-before-group? (:group? prev-before))))))
+
+(defn generate-move-token-set
+  "Create changes for dropping a token set or token set.
+  Throws for impossible moves."
+  [changes tokens-lib drop-opts]
+  (when-let [drop-opts' (calculate-move-token-set-or-set-group tokens-lib drop-opts)]
```

Comment:
All `generate-xx` functions should always return a changes object, that may be the same as the input one. The idea is that you can chain several generates together.

Please change the `when-let` by an `if-let` that returns changes in the else part (and update the test).